source/2_platform/rex_directx/src/directx_renderer.cpp

```diff
@@ -0,0 +1,122 @@
+#include "rex_directx/directx_feature_level.h"
+#include "rex_directx/directx_feature_shader_model.h"
+#include "rex_directx/directx_util.h"
+#include "rex_directx/dxgi/adapter.h"
+#include "rex_directx/dxgi/adapter_manager.h"
+#include "rex_directx/dxgi/factory.h"
+#include "rex_directx/log.h"
+#include "rex_renderer_core/highest_vram_gpu_scorer.h"
+#include "rex_renderer_core/renderer.h"
+#include "rex_std/memory.h"
+
+namespace rex
+{
+  namespace renderer
+  {
+    namespace directx
+    {
+      RendererInfo g_renderer_info; // NOLINT (fuchsia-statically-constructed-objects, cppcoreguidelines-avoid-non-const-global-variables)
+    }                               // namespace directx
+
+    //-------------------------------------------------------------------------
+    const RendererInfo& info()
+    {
+      return directx::g_renderer_info;
+    }
+
+    //-------------------------------------------------------------------------
+    ShaderPlatform shader_platform()
+    {
+      return ShaderPlatform::HLSL;
+    }
+
+    //-------------------------------------------------------------------------
+    bool is_y_up()
+    {
+      return false;
+    }
+
+    //-------------------------------------------------------------------------
+    bool is_depth_0_to_1()
+    {
+      return true;
+    }
+
+    namespace backend
+    {
+      struct DirectXContext
+      {
+        wrl::com_ptr<ID3D12Device> device = nullptr;
+      };
+
+      DirectXContext g_ctx; // NOLINT(fuchsia-statically-constructed-objects, cppcoreguidelines-avoid-non-const-global-variables)
+
+      //-------------------------------------------------------------------------
+      bool initialize()
+      {
+        rsl::unique_ptr<dxgi::Factory> factory = dxgi::Factory::create();
+
+        dxgi::AdapterManager adapter_manager(factory.get(), HighestVramGpuScorer());
+        const dxgi::Adapter* adapter = static_cast<const dxgi::Adapter*>(adapter_manager.selected()); // NNOLINT(cppcoreguidelines-pro-type-static-cast-downcast)
```

Comment:
please look at the comments I've given regarding the Gpu design.
I think as soon as you have to down cast to a type, there's something wrong with the code design.
in this case the down casting can be prevented